### PR TITLE
Fls dls field masking tests

### PR DESCRIPTION
**Pull request by kt-eliatra**
Signed-off-by: Kacper Trochimiak [kacper.trochimiak@eliatra.com](mailto:kacper.trochimiak@eliatra.com)

### Description
[Describe what this change achieves]
* Category (Integration tests)
* Why these changes are required?
* What is the old behavior before changes and new behavior after changes?

Integration tests for FLS, DLS and field masking added.

### Issues Resolved
[List any issues this PR will resolve]

Is this a backport? If so, please add backport PR # and/or commits #

### Testing
[Please provide details of testing done: unit testing, integration testing and manual testing]

### Check List
- [ ] New functionality includes testing
- [ ] New functionality has been documented
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java [hunk 1]
```diff
@@ -0,0 +1,1113 @@
...
+
+	static final TreeMap<String, Song> SECOND_INDEX_SONGS_BY_ID = new TreeMap<>() {{
+		put(SECOND_INDEX_ID_SONG_1, SONGS[3]);
+		put(SECOND_INDEX_ID_SONG_2, SONGS[2]);
+		put(SECOND_INDEX_ID_SONG_3, SONGS[1]);
+		put(SECOND_INDEX_ID_SONG_4, SONGS[0]);
+	}};
+
+	@BeforeClass
+	public static void createTestData() {
```
Comment (peternied) [inline src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java:305]: These tests are very complex to setup, and the tests themselves are verifying many different behaviors - its hard to judge the correctness of a test as written.

Can we isolate the FLS testing into different blocks, or instead orient them so they are more cleanly verified.  Diving into the exact request/response isn't needed, but making sure we the field is present or not is the test.  Using functions to abstract these stages of the test will make them approachable

```java
public void testFlsEnabledFieldsAreHiddenForNormalUsers() {
   // setup
   //   Create Index A, with Document B1, B2, with fields C1, C2, C3
   //    For the test user enforce FLS on C2
   // action & verify
   //   Search query plain - verify C2 isn't visible
   //   Search query aggregation - verify C2 isn't visible
   //   ...
}
```

**Review by peternied — CHANGES_REQUESTED** [review]
Thank you for the exhaustive tests, I know this had to take considerable work to author.  To get this ready to merge I would like to see the tests easier to inspect and update.

The current structure is daunting and I would be concerned that tests themselves are hard to keep updated with the functionality - I've left more specific comments, what you think about adding more abstractions to the test scenarios?

File: src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java (hunk 1 above)
Comment (kt-eliatra) [reply src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java:266]: Done: https://github.com/opensearch-project/security/pull/2258/commits/8579eabed13d2722f2aaf83d8e89cce50ea54a4c

File: src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java [hunk 2]
```diff
@@ -0,0 +1,953 @@
...
+			searchRequest = new SearchRequest("*".concat(indexName));
+			searchResponse = restHighLevelClient.search(searchRequest, DEFAULT);
+
+			assertSearchHitsDoNotContainField(searchResponse, FIELD_STARS);
+
+			//search via alias
+			searchRequest = new SearchRequest(indexAlias);
+			searchResponse = restHighLevelClient.search(searchRequest, DEFAULT);
+
+			assertSearchHitsDoNotContainField(searchResponse, FIELD_STARS);
```
Comment (peternied) [inline src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java:334]: I think these reuses of request/response are a good case for a loop, can we generate a list of request and then iterate through them repeating the same verification?


Maybe something like the following would be easy to generate all the index patterns?  Alternatively if `new SearchRequest(...)` is easy as a one liner that would be easy to read/maintain
```java
   List<String> indicesToCheck = List.of(
      "*".concat(indexName),
      indexAlias,
      ...);
   // Loop over request/response
```

Comment (kt-eliatra) [reply src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java]: fixed

File: src/integrationTest/java/org/opensearch/security/CrossClusterSearchTests.java [hunk 3]
```diff
@@ -155,14 +155,14 @@ public CrossClusterSearchTests(Boolean ccsMinimizeRoundtrips) {
 	@BeforeClass
 	public static void createTestData() {
 		try(Client client = remoteCluster.getInternalNodeClient()){
-			client.prepareIndex(SONG_INDEX_NAME).setId(SONG_ID_1R).setRefreshPolicy(IMMEDIATE).setSource(SONGS[0]).get();
-			client.prepareIndex(SONG_INDEX_NAME).setId(SONG_ID_6R).setRefreshPolicy(IMMEDIATE).setSource(SONGS[5]).get();
-			client.prepareIndex(PROHIBITED_SONG_INDEX_NAME).setId(SONG_ID_3R).setRefreshPolicy(IMMEDIATE).setSource(SONGS[1]).get();
-			client.prepareIndex(LIMITED_USER_INDEX_NAME).setId(SONG_ID_5R).setRefreshPolicy(IMMEDIATE).setSource(SONGS[4]).get();
+			client.prepareIndex(SONG_INDEX_NAME).setId(SONG_ID_1R).setRefreshPolicy(IMMEDIATE).setSource(SONGS[0].asMap()).get();
```
Comment (DarshitChanpura) [inline src/integrationTest/java/org/opensearch/security/CrossClusterSearchTests.java:158]: out of curiosity, why this change? I see similar changed below too and am wondering if it has something to do with structure of query response.

File: src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java [hunk 4]
```diff
@@ -0,0 +1,946 @@
...
+			String aggregationName = "averageStars";
+			SearchRequest searchRequest = averageAggregationRequest(FIRST_INDEX_NAME, aggregationName, FIELD_STARS);
+
+			SearchResponse searchResponse = restHighLevelClient.search(searchRequest, DEFAULT);
+
+			assertThat(searchResponse, isSuccessfulSearchResponse());
+			assertThat(searchResponse, containAggregationWithNameAndType(aggregationName, "avg"));
+			Aggregation actualAggregation = searchResponse.getAggregations().get(aggregationName);
+			assertThat(actualAggregation, instanceOf(ParsedAvg.class));
+			assertThat(((ParsedAvg) actualAggregation).getValue(), not(Double.POSITIVE_INFINITY));
```
Comment (DarshitChanpura) [inline src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java:732]: what is the expected value here?

File: src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java [hunk 5]
```diff
@@ -0,0 +1,946 @@
...
+			assertThat(response, numberOfSearchItemResponsesIsEqualTo(2));
+
+			MultiSearchResponse.Item[] responses = response.getResponses();
+
+			assertThat(responses[0].getResponse(), searchHitsContainDocumentWithId(0, FIRST_INDEX_NAME, firstSongId));
+			assertThat(responses[0].getResponse(), searchHitContainsFieldWithValue(0, FIELD_ARTIST, ARTIST_TWINS));
+			assertThat(responses[1].getResponse(), searchHitsContainDocumentWithId(0, SECOND_INDEX_NAME, secondSongId));
+			assertThat(responses[1].getResponse(), searchHitContainsFieldWithValue(0, FIELD_ARTIST, ARTIST_FIRST));
+		}
+	}
```
Comment (DarshitChanpura) [inline src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java:917]: I have a question, should there be tests about multi-get or multi-search via Aliases. I'm not sure if this scenario is supported or not, but if it is then we should add these tests

**Review by DarshitChanpura — COMMENTED** [review]
Thank you for adding these detailed tests.

File: src/integrationTest/java/org/opensearch/security/CrossClusterSearchTests.java (hunk 3 above)
Comment (kt-eliatra) [reply src/integrationTest/java/org/opensearch/security/CrossClusterSearchTests.java:158]: In order not to lose information about the types of fields that we have in the Song class, I changed the type of the Song.SONGS variable from Map[] to Song[]. I cannot pass song to the IndexRequestBuilder setSource(Object... source) method, it expects an array containing pairs of field name and value. I need to call song.asMap().

File: src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java (hunk 4 above)
Comment (kt-eliatra) [reply src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java]: User has access to the stars field, so the average value should be calculated. I've changed the assertion: https://github.com/opensearch-project/security/pull/2258/commits/e882b4a79e2657b6dd4c1a173fef2c4154b056b7

File: src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java (hunk 5 above)
Comment (kt-eliatra) [reply src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java:946]: Done: https://github.com/opensearch-project/security/pull/2258/commits/e839ba8793a9db38bf8adf9f2968466455082e6d

File: src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java (hunk 2 above)
Comment (peternied) [reply src/integrationTest/java/org/opensearch/security/FlsDlsAndFieldMaskingTest.java]: Thank you for making this update, much cleaner - can we do this to all the tests in this PR?

**Review by peternied — DISMISSED** [review]
Thanks for iterating on this, much more consumable! Thanks

**Comment by peternied** [comment]
@kt-eliatra Please rebase/merge from main this should resolve the CI issues so this change can be merged. Note; I tried to update this PR myself, but I could not push it to your branch.

**Comment by kt-eliatra** [comment]
> @kt-eliatra Please rebase/merge from main this should resolve the CI issues so this change can be merged. Note; I tried to update this PR myself, but I could not push it to your branch.

Done, unfortunately one of the checks was not successful.

**Comment by scrawfor99** [comment]
@cwperks @DarshitChanpura @RyanL1997 can one of you try rerunning the failed backwards compatibility test here? It seems likely it was just a flake-y test.

**Review by cwperks — APPROVED** [review]

**Review by DarshitChanpura — APPROVED** [review]
